File: mobile/android/core/auth/src/main/kotlin/com/slopssaloon/omen/core/auth/AuthRepositorySessionRefresher.kt
package com.slopssaloon.omen.core.auth

import com.slopssaloon.omen.core.session.SessionRefreshOutcome
import com.slopssaloon.omen.core.session.SessionRefreshing

/**
 * Production [SessionRefreshing], over [AuthRepository.refresh].
 *
 * iOS mirror: `App/Api/AuthRepositorySessionRefresher.swift`. Serialization of concurrent
 * refreshes lives in `SessionManager` (a `Mutex`) rather than here, so a caller that queues
 * behind an in-flight refresh can take the freshly stored token instead of starting a second
 * round trip against a refresh token Supabase has already rotated.
 */
class AuthRepositorySessionRefresher(
    private val repository: AuthRepository,
) : SessionRefreshing {
    override suspend fun refreshedSession(): SessionRefreshOutcome =
        when (val outcome = repository.refresh()) {
            is AuthOutcome.Success -> SessionRefreshOutcome.Renewed(outcome.session)
            // Transport-level only. The refresh token was never judged, so neither is the user.
            is AuthOutcome.RetryableError -> SessionRefreshOutcome.Unavailable
            else -> SessionRefreshOutcome.Rejected
        }
}

File: mobile/ios/OmenIOS/OmenIOS/App/Api/AuthRepositorySessionRefresher.swift
import Foundation

/// Production `SessionRefreshing`, over `AuthRepository.refresh()`.
///
/// Its whole job beyond the mapping is **coalescing**. The Command Center fires summary,
/// standings, and moves concurrently; without a shared in-flight task an expired token would
/// start three refresh round trips at once. Supabase rotates the refresh token on every
/// successful refresh, so two of those three would present a token the server had already
/// retired — and a user whose session was perfectly fine would be signed out by their own
/// dashboard loading.
@MainActor
final class AuthRepositorySessionRefresher: SessionRefreshing {
    private let repository: AuthRepository
    private var inFlight: Task<SessionRefreshOutcome, Never>?

    init(repository: AuthRepository) {
        self.repository = repository
    }

    func refreshedSession() async -> SessionRefreshOutcome {
        if let inFlight { return await inFlight.value }

        let task = Task<SessionRefreshOutcome, Never> { [repository] in
            switch await repository.refresh() {
            case .success(let session):
                return .renewed(session)
            case .retryableError:
                // Transport-level only. The refresh token was never judged, so neither is
                // the user — see `SessionRefreshOutcome`.
                return .unavailable
            default:
                return .rejected
            }
        }
        inFlight = task
        let outcome = await task.value
        inFlight = nil
        return outcome
    }
}
